Firestore/core/src/firebase/firestore/immutable/append_only_list.h

```diff
@@ -0,0 +1,228 @@
+
+    new_contents->push_back(value);
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the given value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList push_back(T&& value) const {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->push_back(std::move(value));
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList constructing a new value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  template <typename... Args>
+  ABSL_MUST_USE_RESULT AppendOnlyList emplace_back(Args&&... args) {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->emplace_back(std::forward<Args>(args)...);
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the final link in the chain removed.
+   *
+   * Note that the element isn't actually removed from the backing vector and
+   * it still constitutes the end of the chain. This means that any `push_back`
+   * on the resulting AppendOnlyList will result in a full copy.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList pop_back() const {
+    if (size_ <= 1) {
+      return clear();
+    }
+
+    return AppendOnlyList(contents_, size_ - 1);
+  }
+
+  /**
+   * Creates a new AppendOnlyList without any elements.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList clear() const {
+    return AppendOnlyList(nullptr, 0);
+  }
+
+  size_t size() const {
+    return size_;
+  }
+
+  bool empty() const {
+    return size_ == 0;
+  }
+
+  const_iterator begin() const {
+    if (size_ == 0) {
+      return nullptr;
+    } else {
+      return contents_->data();
+    }
+  }
+
+  const_iterator end() const {
+    if (size_ == 0) {
+      return nullptr;
+    } else {
+      return contents_->data() + size_;
+    }
+  }
+
+  const T& front() const {
+    return *begin();
+  }
+
+  const T& back() const {
+    const T* address = contents_->data() + size_ - 1;
+    return *address;
+  }
+
+  const T& operator[](size_t pos) const {
+    const T* address = contents_->data() + pos;
+    return *address;
+  }
+
+  friend bool operator==(const AppendOnlyList& lhs, const AppendOnlyList& rhs) {
+    return absl::c_equal(lhs, rhs);
+  }
+
+  friend bool operator!=(const AppendOnlyList& lhs, const AppendOnlyList& rhs) {
+    return !(lhs == rhs);
+  }
+
+ private:
+  AppendOnlyList(std::shared_ptr<std::vector<T>> contents, size_t size)
+      : contents_(std::move(contents)), size_(size) {
+  }
+
+  std::shared_ptr<std::vector<T>> PrepareForAppend(size_t new_size) const {
+    std::shared_ptr<std::vector<T>> new_contents;
+
+    if (contents_ && contents_->size() == size_) {
```

Comment:
Your new version causes two allocations in the second branch. The vector first allocates `size()` elements, then `reserve()` triggers a reallocation.

Rephrasing like this would save the allocation:
```c++
auto new_contents = std::make_shared<std::vector<T>>();
new_contents->reserve(new_size);
new_contents->insert(new_contents->begin(), begin(), end());
return new_contents;
```
I've changed this around though in order to avoid invalidation so the committed version is different.